tenders/hvt/hvt_cpu_aarch64.c

```diff
@@ -31,39 +31,107 @@
 
 #include "hvt_cpu_aarch64.h"
 
+/*
+ * If we want support memory greater than 512GB, we should
+ * modify the following PAGE_TABLE_OFFSET at the same time.
+ */
+
 /* First page in AARCH64_PAGE_TABLE is used for PGD */
 #define PGT_PGD_START   0
+
 /* Second page in AARCH64_PAGE_TABLE is used for PUD */
 #define PGT_PUD_START   1
 
-#define PGT_PAGE(x)   (AARCH64_PAGE_TABLE + PAGE_SIZE * (x))
+/* Third page in AARCH64_PAGE_TABLE is used for PMD */
+#define PGT_PMD_START   2
 
-static uint64_t aarch64_mapping_virt_to_phys(uint8_t *va_addr,
-                                uint64_t start, uint64_t size)
-{
-    struct pud *pud;
-    uint32_t idx;
-    uint64_t phys_pud, out_address;
+/* Fourth page in AARCH64_PAGE_TABLE is used for PTE */
+#define PGT_PTE_START   3
 
-    /* The block size of each PUD entry is 1GB */
-    size = PUD_SIZE * DIV_ROUND_UP(size, PUD_SIZE);
+#define PGT_PAGE_ADDR(x)   (AARCH64_PAGE_TABLE + PAGE_SIZE * (x))
 
-    /* Locate the page of PUD to do this translation */
-    phys_pud = PGT_PAGE(PGT_PUD_START + DIV_ROUND_UP(start, PGD_SIZE));
-    pud = (struct pud*)(va_addr + phys_pud);
+enum PGT_LEVEL {
+    L1_TABLE = 1, // PUD
+    L2_TABLE = 2, // PMD
+    L3_TABLE = 3, // PTE
+};
 
-    /* Fill PUD entries */
+static uint64_t aarch64_create_pagetable(enum PGT_LEVEL level,
+                        uint8_t *va_addr, uint64_t start, uint64_t size,
+                        uint64_t attrs, uint64_t attrs_io)
+{
+    uint32_t idx;
+    uint64_t *pentry;
+    uint64_t phys_table_addr, block_size, out_address;
+
+    if (level == L1_TABLE) {
+        /* The block size of each PUD entry is 1GB */
```

Comment:
The comments here are unnecessary and repetitive. I'd prefer that you remove them entirely, or add a suitable generic comment at the start of the if / else block explaining what it does.